include/flang/eval/value.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <flang/parse/ast.hpp>
+#include <functional>
+#include <list>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace flang {
+
+class NullValue;
+class IntegerValue;
+class BuiltinFuncValue;
+class BoolValue;
+class ListValue;
+class UserFuncValue;
+
+using Value = std::variant<NullValue, IntegerValue, BuiltinFuncValue, BoolValue, ListValue, UserFuncValue>;
+
+// TODO: probably should be a singleton
+class NullValue {
+ public:
+  explicit NullValue() {}
+};
+
+class IntegerValue {
+ public:
+  explicit IntegerValue(int value) : _value(value) {}
+
+  int value() const { return _value; }
+
+ private:
+  int _value;
+};
+
+class BoolValue {
+ public:
+  explicit BoolValue(bool value) : _value(value) {}
+
+  bool value() const { return _value; }
+
+ private:
+  bool _value;
+};
+
+class ListValue {
+ public:
+  explicit ListValue(std::list<Value> list) : _value(std::make_shared<std::list<Value>>(list)) {}
+
+  std::shared_ptr<std::list<Value>> value() const { return _value; }
+
+ private:
+  // 🤓 Haha we cannot track cyclic links.
```

Comment:
1) Do you really need list here? What for?
2) Why you use shared ptr here? I think there is no need in smart pointers in this place